### DISCUSSION TITLE
i3 sluggish with more windows?

### DISCUSSION THREAD

**Discussion by eater** [i3/i3#4521] [Q&A]
I've got a decent laptop (i7-7600U CPU @ 2.80GHz) running i3 4.19.1. 

I'd like to use i3 to manage my browser windows, say 30 pages open in Chromium or Firefox. If they're all browser tabs within one browser window, everything's super fast. But if I open them as 30 browser windows in i3 tabs, the system becomes noticeably, annoyingly slow and hot.

The only other WM i've tried is XFCE's, and it does not seem to slow down under the same conditions.

Any tips for speeding it up?

**Comment by orestisfl** [comment]
Does the cpu time of i3 significantly increase under these circumstances? Check top/htop

**Comment by eater** [comment]
After playing around with it a little more, I realize that the slowness is only apparent when the workspace with all the tabbed windows is in view. Change workspace, it goes back to normal.

The `i3` process stays around 0.0% CPU and memory! It's the browser processes that are at the top of `top`.

**Reply by michaelmrose** [reply]
I strongly suspect the culprit is the browser not the window manager here. Does chrome allocate more resources for tabs which are the focused tab for their window?  Alternatively is chrome unable to tell which are visible in a tabbed layout and lets tabs which are visible do more work than background tabs? 

Honestly although it would be efficient to use one method of tabbing i3's tabs are strictly inferior to application tabs in that it lacks the context of the application.  You can close groups of tabs, bookmark groups of tabs, save them as a session for recall later. Choose between opening a particular link in the background or foreground. You can close all tabs save for a particular grouping.

This isn't a lack in i3 its just never going to have that app context. I3's facility to tab many unalike windows is great but its not a replacement for application tabs.

**Reply by Airblader** [reply]
If you are running a compositor, you could try disabling that to see if it changes anything. There's a chance this is somehow related to `_NET_WM_STATE_HIDDEN`, browsers have reacted in funny ways to that in the past already.

**Reply by eater** [reply]
I am not running a compositor, @Airblader, thanks. 

I don't think it's solely the browsers that are to blame, because I don't get the same slowdown when I have all the same browser windows open on an XFCE workspace.

**Reply by Airblader** [reply]
The fact that it works on other window managers disproves that it's the browser fault just as much as other applications working in i3 disproves that it's i3's fault. So yes, we can't be quick here to say who is to blame, it's still fair game because we don't know what's going on. A browser behaving "badly" could well be due to incorrect behavior on the i3 side, indeed.

**Comment by alexeymetelev** [comment]
Chrome has "Tab throttling" feature, see https://blog.chromium.org/2020/11/tab-throttling-and-more-performance.html. This helps to save CPU usage and battery time.

When you open many browser instances on different i3wm workspaces, active tabs on all instances are working without throttling even when non of the browsers are visible. 

You can check it by running this code in the Chrome Dev Tool Console. It is running faster when tab is active and slower when it is not active.

```javascript
let chainCount = 0;

setInterval(() => {
  chainCount++;
  console.log(`This is number ${chainCount} in the chain`);
}, 500);
```

Also it is running slower (throttling is on) when window has _NET_WM_STATE_HIDDEN.
You can check it by setting this propery forcibly by using this commands in your terminal:
```bash
# get window id
sleep 5 && xdotool getwindowfocus

# set property
xprop -f _NET_WM_STATE 32a -id 71303313 -set _NET_WM_STATE _NET_WM_STATE_HIDDEN

# get all properties
xprop -id #{window id}
```

I would suggest to set _NET_WM_STATE_HIDDEN for all windows which are not visible to user.

**Reply by Airblader** [reply]
> I would suggest to set _NET_WM_STATE_HIDDEN for all windows which are not visible to user.

i3 already sets this on according to spec.

**Reply by vincentbernat** [reply]
For me, it does not work either:

```
 10:54 ❱ xprop -id 0x200002c | grep -E '_NET_WM_(STATE|NAME)'
_NET_WM_STATE(ATOM) =
_NET_WM_NAME(UTF8_STRING) = "CrystalBlaster C2 - Serdashop — Mozilla Firefox"
```

But maybe, I am not selecting the right X11 window.

**Reply by Airblader** [reply]
I said it sets it according to spec, not that it sets it on all windows. The spec says

> _NET_WM_STATE_HIDDEN should be set by the Window Manager to indicate that a window would not be visible on the screen if its desktop/viewport were active and its coordinates were within the screen bounds.

Which is why we set that hint on the non-active tabs of stacked / tabbed containers. If you eg use i3 tabs instead of browser tabs, most of them would have that hint set.

**Reply by vincentbernat** [reply]
It becomes hidden only if I hide it with tab or stack layout. The workspace not being displayed does not seem to be enough.

```c
bool con_is_hidden(Con *con) {
    Con *current = con;

    /* ascend to the workspace level and memorize the highest-up container
     * which is stacked or tabbed. */
    while (current != NULL && current->type != CT_WORKSPACE) {
        Con *parent = current->parent;
        if (parent != NULL && (parent->layout == L_TABBED || parent->layout == L_STACKED)) {
            if (TAILQ_FIRST(&(parent->focus_head)) != current)
                return true;
        }

        current = parent;
    }

    return false;
}
```

**Reply by vincentbernat** [reply]
> The spec says [...]

Ack!
